src/data_structures/avl_tree.js

```diff
@@ -0,0 +1,528 @@
+  this.postOrder(current.right, callback);
+};
+
+/**
+ * Pre-order traversal from the given node.
+ */
+AVLTree.prototype.preOrder = function(current, callback) {
+  if (!current) {
+    return;
+  }
+  if (typeof callback === 'function') {
+    callback(current);
+  }
+  this.preOrder(current.left, callback);
+  this.preOrder(current.right, callback);
+};
+
+/**
+ * Finds a node by its value.
+ */
+AVLTree.prototype.find = function(value) {
+  return this._find(value, this.root);
+};
+
+/**
+ * Finds a node by its value in the given sub-tree.
+ */
+AVLTree.prototype._find = function(value, current) {
+  if (!current) {
+    return null;
+  }
+
+  if (current.value === value) {
+    return current;
+  }
+  if (current.value > value) {
+    return this._find(value, current.left);
+  }
+  if (current.value < value) {
+    return this._find(value, current.right);
+  }
+};
+
+/**
+ * Replaces the given child with the new one,
+ * for the given parent.
+ */
+AVLTree.prototype.replaceChild = function(parent, oldChild, newChild) {
+  if (parent === null) {
+    this.root = newChild;
+    if (this.root !== null) {
+      this.root.parent = null;
+    }
+  } else {
+    if (parent.left === oldChild) {
+      parent.left = newChild;
+    } else {
+      parent.right = newChild;
+    }
+    if (newChild) {
+      newChild.parent = parent;
+    }
+  }
+};
+
+/**
+ * Removes a node by its value.
+ */
+AVLTree.prototype.remove = function(value) {
+  var node = this.find(value);
+  if (!node) {
+    return false;
+  }
+
+  if (node.left && node.right) {
+    var min = this.findMin(node.right);
+    var temp = node.value;
+    node.value = min.value;
+    min.value = temp;
+    return this.remove(min);
+  } else {
+    if (node.left) {
+      this.replaceChild(node.parent, node, node.left);
+      this.keepHeightBalance(node.left, true);
+    } else if (node.right) {
+      this.replaceChild(node.parent, node, node.right);
+      this.keepHeightBalance(node.right, true);
+    } else {
+      this.replaceChild(node.parent, node, null);
+      this.keepHeightBalance(node.parent, true);
+    }
+    return true;
+  }
+};
+
+/**
+ * Finds the node with minimum value in the given
+ * sub-tree.
+ */
+AVLTree.prototype._findMin = function(node, current) {
+  current = current || {
+    value: Infinity
+  };
+  if (!node) {
+    return current;
+  }
+  if (current.value > node.value) {
+    current = node;
+  }
+  return this._findMin(node.left, current);
+};
+
+/**
+ * Finds the node with maximum value in the given
+ * sub-tree.
+ */
+AVLTree.prototype._findMax = function(node, current) {
+  current = current || {
+    value: -Infinity
+  };
+  if (!node) {
+    return current;
+  }
+  if (current.value < node.value) {
+    current = node;
+  }
+  return this._findMax(node.right, current);
+};
+
+/**
+ * Finds the node with minimum value in the whole tree.
+ */
+AVLTree.prototype.findMin = function() {
+  return this._findMin(this.root);
```

Comment:
_findMin   and findMin    what's the best about this XD  additional method thing